### PR TITLE
[SPARK] Allow existing spark2 JMH benchmarks to work with either spark 2 or spark 3

### PR DESCRIPTION
**Pull request by kbendick**
Right now, we have JMH benchmarks to allow performance comparisons between the native spark sources / sinks vs the iceberg sources and sinks in spark.

However, these JMH test suites don't get registered for Spark 3.

I have moved the JMH test code to `spark/src/jmh/...` and then added that to the src directories for those projects.

Now, depending on which JDK one is using, you can run either Spark 2 or 3 tests (with JDK 8) or Spark 3 tests (with JDK 11).

It would still be possible to add specific benchmarks to Spark 3 or Spark 2 by placing them in `spark[2|3]/src/jmh/...`.

For now, all of the code is left as is and has been copied over (with the command to run updated in each file). I've also added a `.gitkeep` file in `spark2/benchmark` and `spark3/benchmark` so that users can run the commands as is without having to create the folders.

This closes https://github.com/apache/iceberg/issues/2590

### CODE REVIEW

**Comment by kbendick** [comment]
cc @aokolnychyi @RussellSpitzer @flyrain @karuppayya @raptond @rdblue

File: jmh.gradle [hunk 1]
```diff
@@ -17,7 +17,14 @@
  * under the License.
  */
 
-def jmhProjects = [ project("iceberg-spark2") ]
+def jmhProjects = []
+if (JavaVersion.current() == JavaVersion.VERSION_1_8) {
+  jmhProjects = [ project("iceberg-spark2") ]
```
Comment (holdenk) [inline jmh.gradle:22]: I _think_ Spark 3 supports JDK 1.8, would it make sense to do iceberg-spark3 here as well? Or is there a reason for only doing one of them?

**Review by holdenk — COMMENTED** [review]
One quick question, but thanks for working on this :)

File: jmh.gradle (hunk 1 above)
Comment (kbendick) [reply jmh.gradle:22]: It would indeed make sense.

I was having trouble getting both projects to work at one time, so I chose to just skip it as that's about as far as my gradle abilities got me. I tried adding to the array, etc, but the plugin didn't like that.

I'll look into it further to see if it can be supported. Thanks for the review!

Comment (kbendick) [reply jmh.gradle:22]: Looking into it again, I think probably the issue was that I didn't run `./gradlew clean` when switching between JDKs locally. Pushing a fix now!

Comment (kbendick) [reply jmh.gradle:22]: Ah. Confirmed the issue is 100% with my own understanding of groovy / gradle (and always wanting it to be more like Scala).

Comment (kbendick) [reply jmh.gradle]: Alright. This has been fixed :). Confirmed by setting my terminals JDK to java 8, then running both a spark 2 and spark 3 JMH test, then setting to Java 11, and running a Spark 3 test (and back again). 👍

File: jmh.gradle [hunk 2]
```diff
@@ -17,7 +17,14 @@
  * under the License.
  */
 
-def jmhProjects = [ project("iceberg-spark2") ]
+def jmhProjects
```
Comment (RussellSpitzer) [inline jmh.gradle:20]: The formatting here is a bit odd but other than this, looks good to me

**Comment by RussellSpitzer** [comment]
@kbendick  got this error
```* What went wrong:
Execution failed for task ':iceberg-spark2:jmh'.
> A failure occurred while executing me.champeau.gradle.IsolatedRunner
   > Error during execution of benchmarks
   ```
   
   When i ran 
   ```g jmh```
   
   on the pr

File: jmh.gradle (hunk 2 above)
Comment (kbendick) [reply jmh.gradle:20]: Yeah. I could use `deef jmhProjects = []`, but IntelliJ complains about unused variable.

I then tried to be a bit more Scala like, and did `def jmhProjects = { ... return array_of_what_i_want }` but that generated a closure.

Basically, the best I think I can offer is `def jmhProjects = []` and then leave it like that.

**Comment by kbendick** [comment]
> @kbendick got this error
> 
> ```
> Execution failed for task ':iceberg-spark2:jmh'.
> > A failure occurred while executing me.champeau.gradle.IsolatedRunner
>    > Error during execution of benchmarks
> ```
> 
> When i ran
> `g jmh`
> 
> on the pr

So just like the current set up (as far as I'm aware), you have to specify the actual test to be run. So something like the following, as mentioned in the existing code [here](https://github.com/apache/iceberg/pull/2595/files#diff-ed844daf9dae5835b7ed247c21439af720df778c28596c644a9da8d7013f4fb9L56-L58)
```
./gradlew :iceberg-spark2:jmh \ 
   -PjmhIncludeRegex=SparkParquetWritersNestedDataBenchmark \
   -PjmhOutputPath=benchmark/spark-parquet-writers-nested-data-benchmark-result.txt
```

I will test simply `./gradlew jmh` on the current state of master vs my PR, but with my PR, the tests run the same way as mentioned in the doc comments of each test. @RussellSpitzer

**Comment by kbendick** [comment]
> > @kbendick got this error
> > ```
> > Execution failed for task ':iceberg-spark2:jmh'.
> > > A failure occurred while executing me.champeau.gradle.IsolatedRunner
> >    > Error during execution of benchmarks
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > When i ran
> > `g jmh`
> > on the pr
> 
> So just like the current set up (as far as I'm aware), you have to specify the actual test to be run. So something like the following, as mentioned in the existing code [here](https://github.com/apache/iceberg/pull/2595/files#diff-ed844daf9dae5835b7ed247c21439af720df778c28596c644a9da8d7013f4fb9L56-L58)
> 
> ```
> ./gradlew :iceberg-spark2:jmh \ 
>    -PjmhIncludeRegex=SparkParquetWritersNestedDataBenchmark \
>    -PjmhOutputPath=benchmark/spark-parquet-writers-nested-data-benchmark-result.txt
> ```
> 
> I will test simply `./gradlew jmh` on the current state of master vs my PR, but with my PR, the tests run the same way as mentioned in the doc comments of each test. @RussellSpitzer

I tried running `./gradlew jmh` both from master and from my branch. Both of them failed eventually, but after running for 20-25 minutes and completing at least 3-4 of the benchmarks each time. @RussellSpitzer has now been able to get plain `./gradle jmh` to run for over an hour now on this PR branch.

When I run an individual test, as recommended in the comments of each test (or above), it runs just fine to completion.

I think it's more an issue with gradle JVM settings and then running out of memory or other resources, possibly with the number of forked threads when running all of the benchmarks for all of the requested iterations.

**TLDR:** I don't think we should run `jmh` itself to run _all_ tests without better configuring gradle. But stick to running the tests as they're mentioned in the test comments. This does not introduce any new issues with the JMH tests, other than possibly the added number of tests if one runs `./gradlew jmh` by itself (which is not guaranteed to run to completion anyway with the current state in master).

File: jmh.gradle (hunk 2 above)
Comment (kbendick) [reply jmh.gradle]: Discussed offline and changed it to be initialized with an array for readability.

**Comment by kbendick** [comment]
> > > @kbendick got this error
> > > ```
> > > Execution failed for task ':iceberg-spark2:jmh'.
> > > > A failure occurred while executing me.champeau.gradle.IsolatedRunner
> > >    > Error during execution of benchmarks
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > When i ran
> > > `g jmh`
> > > on the pr
> > 
> > 
> > So just like the current set up (as far as I'm aware), you have to specify the actual test to be run. So something like the following, as mentioned in the existing code [here](https://github.com/apache/iceberg/pull/2595/files#diff-ed844daf9dae5835b7ed247c21439af720df778c28596c644a9da8d7013f4fb9L56-L58)
> > ```
> > ./gradlew :iceberg-spark2:jmh \ 
> >    -PjmhIncludeRegex=SparkParquetWritersNestedDataBenchmark \
> >    -PjmhOutputPath=benchmark/spark-parquet-writers-nested-data-benchmark-result.txt
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I will test simply `./gradlew jmh` on the current state of master vs my PR, but with my PR, the tests run the same way as mentioned in the doc comments of each test. @RussellSpitzer
> 
> I tried running `./gradlew jmh` both from master and from my branch. Both of them failed eventually, but after running for 20-25 minutes and completing at least 3-4 of the benchmarks each time. @RussellSpitzer has now been able to get plain `./gradle jmh` to run for over an hour now on this PR branch.
> 
> When I run an individual test, as recommended in the comments of each test (or above), it runs just fine to completion.
> 
> I think it's more an issue with gradle JVM settings and then running out of memory or other resources, possibly with the number of forked threads when running all of the benchmarks for all of the requested iterations.
> 
> **TLDR:** I don't think we should run `jmh` itself to run _all_ tests without better configuring gradle. But stick to running the tests as they're mentioned in the test comments. This does not introduce any new issues with the JMH tests, other than possibly the added number of tests if one runs `./gradlew jmh` by itself (which is not guaranteed to run to completion anyway with the current state in master).

Further follow up - @RussellSpitzer and I were both able to get the tests to run to completion, but users should not run `~/.gradlew jmh` without modifying their gradle parameters quite significantly. It's best to run tests one at a time, as mentioned in the tests themselves.

This patch _**does not change the behavior of the jmh test suite**_ with respect to ability to run to completion from its previous state. It's still possible either way, just rather hard given the number of tests and the limitations of Gradle's JVM (even with the common expanded memory parameters).

**Comment by flyrain** [comment]
Hi @kbendick , sorry for the delayed review. Both spark2 and spark3 jmh commands works well in my laptop with your PR. For Spark3, I can do with both jdk8 and jdk11. Looks good to me overall. One question, is it possible to let gradle create dirs(`spark2/benchmark` and `spark3/benchmark`) instead of using `.gitkeep` files?

**Review by flyrain — APPROVED** [review]

**Comment by kbendick** [comment]
> Hi @kbendick , sorry for the delayed review. Both spark2 and spark3 jmh commands works well in my laptop with your PR. For Spark3, I can do with both jdk8 and jdk11. Looks good to me overall. One question, is it possible to let gradle create dirs(`spark2/benchmark` and `spark3/benchmark`) instead of using `.gitkeep` files?

I'm admittedly not great with Gradle. I tried to figure out how to get this to work, but couldn't. I could possibly look into it further, but the current state in master is that if a user runs the commands as executed, they get a folder not found error. So I figured this was still an upgrade.

By default, if no output path is specified, the results get placed in a `results.txt` file under the projects `/build/`.

So given that this is a user specified input that does deviate from the defaukf, it might make more sense to simply leave as is (where the user gets an error and then creates the folder). Though I am partial to the gitignore as that ensures nobody updates results from their laptop or some env that might be very resource constrained and not representative.

Let me take another pass at it for a bit, but then maybe we can just revert those changes and leave it with the current behavior. If we really want to have Gradle create the folders, would it be possible to simply remove the `.gitkeep` files and the pre-made folders and then try for that in a follow up PR? Given the age of this PR and its size, I'm somewhat anxious to get it in and then start out any more build file changes in another PR. @flyrain

**Comment by kbendick** [comment]
I spoke to @flyrain offline and he agrees that due to the age of this PR and the complexity it's going to add to the build file (and the subsequent discussion), we should attempt to change the build file to create the output directories in a subsequent PR (as this discussion has already gotten really huge and the build file changes will almost certainly invite more discussion). This is arguably not worse than the current state, though I'm happy to remove the `.gitkeep` folders and keep it in the current state (where specifying the output path causes a failure until the user creates the directory the first time).

**Review by RussellSpitzer — APPROVED** [review]
